Refactors admin PMs to be hopefully more resilient

LemonInTheDark opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request

MSO got me curious, so I refactored admin help code doin my best to follow https://github.com/tgstation/tgstation/pull/68324#issuecomment-1179773970
The basic idea is try and make behavior as clear as possible, and keep data extraction separate from logic
Also fail loud, in this case meaning any operation that would cause the pm to fail to send should send an error message to the sender.

I'm splitting the main proc up into subsections, in an attempt to make it easier to reason about and work with.
As a side effect of this, some operations need to be done more then once to avoid potential debugger escalation.

Related to this, I'm also no longer directly passing around client references through the procs. I instead pass the result of a pretty basic proc into each, to avoid issues with sleeping and such.

Changes the backup for replying to someone who logs out, instead of using the first person to send something to the ticket as our backup, we use the recipient's ckey.

Converts all uses of the linkify span to a macro. One of these doesn't have a closing span, not sure if that's a bug or a feature

Also I just cleaned things up, because the old state of things really bothered me. Too hard to read/understand meaning, and too verbose.
I've also fixed some logic errors, like being unable to respond to tgs ahelps, or a few runtimes in tgs handling.
Did you know secret ckeys, despite claiming to be unique, weren't? 
The hope is this will prevent failure in the tickets system in future

<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->

## Why It's Good For The Game

I really don't want to go back to walking up and down stairs trying to get the sand bug to trigger, guess it starts tomorrow.

## Changelog

<!-- If your PR modifies aspects of the game that can be concretely observed by players or admins you should add a changelog. If your change does NOT meet this description, remove this section. Be sure to properly mark your PRs to prevent unnecessary GBP loss. You can read up on GBP and it's effects on PRs in the tgstation guides for contributors. Please note that maintainers freely reserve the right to remove and add tags should they deem it appropriate. You can attempt to finagle the system all you want, but it's best to shoot for clear communication right off the bat. -->

:cl:
admin: Wooooo spoooky ADMIN HELP BACKEND HAS CHANGED. If you run into any strange behavior or dropped messages, REPORT THEM
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

LemonInTheDark committed:
Begins refactoring admin pms, with a focus on runtime safety and ease of understanding

There's more I want to do here, mostly with spans and cleaning up sone
logic flow to be easier to understand

The primary thrust of it is making explicit any assumptions we make about values
This is important because this is sensitive code, and runtimes are not acceptable, since they both leak information and could lead to dropped ahelps, which is almost worse.

I'm also splitting the proc up into subsections, in an attempt to make it easier to reason about and work with
As a side effect of this, some operations need to be done more then once to avoid potential debugger escalation.

Related to this, I'm also no longer directly passing around client references through the procs. I instead pass the result of a pretty basic proc into each, to avoid issues with sleeping and such.

Oh, and I fixed a bug that made it impossible to reply to remote ahelp
responses. I hate this place.

LemonInTheDark committed:
Converts all uses of the linkify span to a macro. One of these doesn't have a closing span, not sure if that's a bug or a feature

LemonInTheDark committed:
Cleans up more logic embedded inside data, since this is critical code

Changes the backup for replying to someone who logs out, instead of
using the first person to send something to the ticket as our backup, we
use the recipient's ckey.

LemonInTheDark committed:
some cleanup

LemonInTheDark committed:
Converts a few more things over to the pattern

Reorganizes and deocuments sends_adminpm_message to be less hard to
track/read

LemonInTheDark committed:
Cleans up a hanging proc that needed filtering

LemonInTheDark committed:
Does a passthrough over tgs, makes stealthmin keys actually unique

LemonInTheDark committed:
cleanup, screed at the top of the file

LemonInTheDark committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into walktimed-tickets

LemonInTheDark committed:
cleanup, comments, etc

TrashDoxx commented on code/__DEFINES/span.dm:
```diff
@@ -60,6 +60,7 @@
 #define span_info(str) ("<span class='info'>" + str + "</span>")
 #define span_infoplain(str) ("<span class='infoplain'>" + str + "</span>")
 #define span_interface(str) ("<span class='interface'>" + str + "</span>")
+#define span_linkfiy(str) ("<span class='linkify'>" + str + "</span>")
```
ehh.. linkfiy? isn't that a mistake? here and further in the code.

TrashDoxx replied:
me too

LemonInTheDark committed:
Thanks trash, I can in fact not spell

san7890 commented on code/modules/admin/verbs/adminpm.dm:
```diff
@@ -113,333 +173,591 @@
+	if(!log_message)
+		return
+
+	var/raw_messsage = log_message
```
Typo? Here and subsequent lines.

LemonInTheDark committed:
ok yeah I really can't spell

jlsnow301 commented:
i love that you've used _lad_ and _boi_ as mob var names here, getting tired of thinking up crafty names myself

LemonInTheDark committed:
updated comment to explain how code works

MrsTonedOne reviewed (approved):
Thanks for the insanely fast turn around on this.

san7890 commented on code/modules/admin/verbs/adminpm.dm:
```diff
@@ -25,80 +42,123 @@
-	if(AH?.opening_responders && length(AH.ticket_interactions) == 1)
-		SEND_SOUND(src, sound('sound/machines/buzz-sigh.ogg', volume=30))
-		message_prompt += "\n\n**This ticket is already being responded to by: [english_list(AH.opening_responders)]**"
+	var/messsage = input(src, message_prompt, request) as message|null
```
even more messsage spellcheck here and subsequent lines

LemonInTheDark commented:
> Thanks for the insanely fast turn around on this.

I got bored and it sounded fun, thanks for the interesting problem

LemonInTheDark committed:
messsage -> message spellcheck

LemonInTheDark committed:
Nullchecking prevention

if for whatever reason someone in the future adds code that lummox says can sneakily allow for Logout() to execute between this and the last time client was checked, this could potentially runtime without nullchecking

Co-authored-by: Kylerace <kylerlumpkin1@gmail.com>

Mooshimi commented:
Don't let the bot close this... it's too vital 👀
